vehicle/bluelink/bluelink.go

```diff
@@ -0,0 +1,375 @@
+	}
+
+	return cookieClient, err
+}
+
+func (v *API) setLanguage(cookieClient *util.HTTPHelper) error {
+	headers := map[string]string{
+		"Content-type": "application/json",
+	}
+
+	data := map[string]interface{}{
+		"lang": "en",
+	}
+
+	req, err := v.jsonRequest(http.MethodPost, v.config.URI+v.config.Lang, headers, data)
+	if err == nil {
+		_, err = cookieClient.Request(req)
+	}
+
+	return err
+}
+
+func (v *API) login(cookieClient *util.HTTPHelper) (string, error) {
+	headers := map[string]string{
+		"Content-type": "application/json",
+	}
+
+	data := map[string]interface{}{
+		"email":    v.user,
+		"password": v.password,
+	}
+
+	req, err := v.jsonRequest(http.MethodPost, v.config.URI+v.config.Login, headers, data)
+	if err != nil {
+		return "", err
+	}
+
+	var redirect struct {
+		RedirectURL string `json:"redirectUrl"`
+	}
+
+	var accCode string
+	if _, err = cookieClient.RequestJSON(req, &redirect); err == nil {
+		if parsed, err := url.Parse(redirect.RedirectURL); err == nil {
+			accCode = parsed.Query().Get("code")
+		}
+	}
+
+	return accCode, err
+}
+
+func (v *API) getToken(accCode string) (string, error) {
+	headers := map[string]string{
+		"Authorization": "Basic ZmRjODVjMDAtMGEyZi00YzY0LWJjYjQtMmNmYjE1MDA3MzBhOnNlY3JldA==",
+		"Content-type":  "application/x-www-form-urlencoded",
+		"User-Agent":    "okhttp/3.10.0",
+	}
+
+	data := "grant_type=authorization_code" +
+		"&redirect_uri=https%3A%2F%2Fprd.eu-ccapi.kia.com%3A8080%2Fapi%2Fv1%2Fuser%2Foauth2%2Fredirect" +
+		"&code=" + accCode
+
+	req, err := v.request(http.MethodPost, v.config.URI+v.config.AccessToken, headers, strings.NewReader(data))
+	if err != nil {
+		return "", err
+	}
+
+	var tokens struct {
+		TokenType   string `json:"token_type"`
+		AccessToken string `json:"access_token"`
+	}
+
+	var accToken string
+	if _, err = v.RequestJSON(req, &tokens); err == nil {
+		accToken = fmt.Sprintf("%s %s", tokens.TokenType, tokens.AccessToken)
+	}
+
+	return accToken, err
+}
+
+func (v *API) getVehicles(accToken, did string) (string, error) {
+	headers := map[string]string{
+		"Authorization":       accToken,
+		"ccsp-device-id":      did,
+		"ccsp-application-id": "693a33fa-c117-43f2-ae3b-61a02d24f417",
+		"offset":              "1",
+		"User-Agent":          "okhttp/3.10.0",
+	}
+
+	req, err := v.request(http.MethodGet, v.config.URI+v.config.Vehicles, headers, nil)
+	if err == nil {
+		var resp response
+		if _, err = v.RequestJSON(req, &resp); err == nil {
+			if len(resp.ResMsg.Vehicles) == 1 {
+				return resp.ResMsg.Vehicles[0].VehicleID, nil
+			}
+
+			err = errors.New("couldn't find vehicle")
+		}
+	}
+
+	return "", err
+}
+
+func (v *API) preWakeup(accToken, did, vid string) error {
+	data := map[string]interface{}{
+		"action":   "prewakeup",
+		"deviceId": did,
+	}
+
+	headers := map[string]string{
+		"Authorization":       accToken,
+		"ccsp-device-id":      did,
+		"ccsp-application-id": "693a33fa-c117-43f2-ae3b-61a02d24f417",
```

Comment:
Wäre es nicht sinnvoller, das irgendwie in eine Konstante zu packen? Im aktuellen File wird's schon zwei Mal genutzt. Und ggf. muss man es irgendwann im Lebenszyklus auch noch mal ändern?